src/ros/ROSInterface.cpp

```diff
@@ -0,0 +1,268 @@
+                kvpair.type = iai_knowledge_msgs::KeyValuePair::TYPE_INT;
+                kvpair.value_int = ((Integer32Term*)term.get())->value();
+                break;
+            case TermType::LONG:
+                kvpair.type = iai_knowledge_msgs::KeyValuePair::TYPE_LONG;
+                kvpair.value_long = ((LongTerm*)term.get())->value();
+                break;
+            case TermType::LIST:
+                kvpair.type = iai_knowledge_msgs::KeyValuePair::TYPE_LIST;
+                ss << ((ListTerm*)term.get());
+                kvpair.value_list = ss.str();
+                break;
+            case TermType::VARIABLE:
+                // TDDO: Can a variable be in the answer?
+            case TermType::PREDICATE:
+                break;
+            case TermType::MODAL_OPERATOR:
+                throw AnswerError(*answer, *term);
+                break;
+        }
+        graphAnswer.substitution.push_back(kvpair);
+    }
+    return graphAnswer;
+}
+
+void ROSInterface::executeAskAllCB(const iai_knowledge_msgs::askallGoalConstPtr& goal)
+{
+
+    // Implement your action here
+    FormulaPtr phi(QueryParser::parse(goal->query.queryString));
+
+    FormulaPtr mPhi = applyModality(goal->query, phi);
+    auto query = std::make_shared<ModalQuery>(mPhi, QUERY_FLAG_ALL_SOLUTIONS);
+
+    auto resultStream = kb_.submitQuery(query->formula(), QUERY_FLAG_ALL_SOLUTIONS);
+    auto resultQueue = resultStream->createQueue();
+
+    int numSolutions_ = 0;
+    iai_knowledge_msgs::askallResult result;
+    while(true) {
+        auto nextResult = resultQueue->pop_front();
+
+        if(AnswerStream::isEOS(nextResult)) {
+            break;
+        }
+        else {
+            if (nextResult->substitution()->empty()) {
+                numSolutions_ = 1;
+                break;
+            } else {
+                // Push one answer
+                iai_knowledge_msgs::GraphAnswer answer = createGraphAnswer(nextResult);
+                result.answer.push_back(answer);
+                numSolutions_ += 1;
+                // publish feedback
+                iai_knowledge_msgs::askallFeedback feedback;
+                feedback.numberOfSolutions = numSolutions_;
+                askall_action_server_.publishFeedback(feedback);
+            }
+        }
+    }
+
+    if(numSolutions_ == 0) {
+        result.status = iai_knowledge_msgs::askallResult::FALSE;
+    } else {
+        result.status = iai_knowledge_msgs::askallResult::TRUE;
+    }
+    askall_action_server_.setSucceeded(result);
+}
+
+void ROSInterface::executeAskOneCB(const iai_knowledge_msgs::askoneGoalConstPtr& goal)
+{
+    // Implement your action here
+    FormulaPtr phi(QueryParser::parse(goal->query.queryString));
+
+    FormulaPtr mPhi = applyModality(goal->query, phi);
+    auto query = std::make_shared<ModalQuery>(mPhi, QUERY_FLAG_ONE_SOLUTION);
+
+    auto resultStream = kb_.submitQuery(query->formula(), QUERY_FLAG_ONE_SOLUTION);
```

Comment:
I was rather thinking we use the other submitQuery interface here where you provide a `GraphQuery` with a `ModalityFrame`. then we could rather construct the `ModalityFrame` directly from the ROS message. An advantage would be that the graph query interface takes a shortcut compared to if you submit a modal query